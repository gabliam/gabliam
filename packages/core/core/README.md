[![Build Status][build-image]][build-url]
[![NPM version][npm-image]][npm-url]


# Gabliam core

Gabliam is on top of inversify.
Gabliam help you for register all class in inversify.
Gabilliam is modular with plugins.


# Gabliam phase

## Build phase

1. Load all config file
2. Load all file in scanPath of main
	- All decorators on class add metadata and can register this class
	- if an class has @scanPath, add this folder in queue for load
	- load all plugin
3. Bind classes in DI (inversify)
	- call all plugin.bind
4. load all config classes (order: CoreConfig => PluginConfig => config)
	- call all plugin.config for all instance of config class
5. Build
	- call all plugin.build

## Start phase
/!\ Build phase must be passed

1. call all plugin.start

## Stop phase
/!\ Build and start phases must be passed

1. call all plugin.start

## Detroy phase
/!\ Stop phase must be passed

1. call all plugin.detroy

# Samples
Go to [Samples]



[Samples]: <https://github.com/gabliam/gabliam/tree/master/examples>

# License

  MIT

[build-image]: https://img.shields.io/travis/gabliam/gabliam/master.svg?style=flat-square
[build-url]: https://travis-ci.org/gabliam/gabliam
[npm-image]: https://img.shields.io/npm/v/@gabliam/core.svg?style=flat-square
[npm-url]: https://www.npmjs.com/package/@gabliam/gabliam/core
