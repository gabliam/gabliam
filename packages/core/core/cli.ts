#!/usr/bin/env node
import 'reflect-metadata';
import yargs from 'yargs';
import { StartCommand } from './src/commands/start-command';

yargs
  .usage('Usage $0 <command> [options]')
  .command(new StartCommand())
  .recommendCommands()
  .demandCommand(1)
  .strict()
  .alias('v', 'version')
  .help('h')
  .alias('h', 'help').argv;

require('yargonaut')
  .style('blue')
  .style('yellow', 'required')
  .helpStyle('green')
  .errorsStyle('red');
