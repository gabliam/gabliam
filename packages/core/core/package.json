{
  "name": "@gabliam/core",
  "version": "0.0.0-PLACEHOLDER",
  "description": "Gabliam is on top of inversify",
  "main": "index.js",
  "typings": "index.d.ts",
  "module": "index.js",
  "scripts": {
    "build": "rimraf lib && tsc -p tsconfig.build.json"
  },
  "files": [
    "lib",
    "src",
    "index.js",
    "index.d.ts"
  ],
  "dependencies": {
    "@gabliam/expression": "0.0.0-PLACEHOLDER",
    "@types/joi": "^14.0.1",
    "caller": "^1.0.1",
    "co": "^4.6.0",
    "debug": "^4.1.1",
    "glob": "^7.1.3",
    "graph-data-structure": "^1.8.0",
    "inversify": "^5.0.1",
    "inversify-logger-middleware": "^3.0.0",
    "joi": "^14.3.1",
    "js-yaml": "^3.12.1",
    "lodash": "^4.17.11",
    "shortstop": "^1.0.3",
    "shortstop-handlers": "^1.0.1",
    "tslib": "^1.9.3"
  },
  "devDependencies": {
    "@types/debug": "^4.1.3",
    "@types/fs-extra": "^5.0.4",
    "@types/glob": "^7.1.1",
    "@types/js-yaml": "^3.11.4",
    "@types/lodash": "^4.14.123",
    "fs-extra": "^7.0.1",
    "properties": "^1.2.1",
    "rxjs": "^6.3.3",
    "toml": "^2.3.3"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/gabliam/gabliam.git"
  },
  "author": "David Touzet <dtouzet@gmail.com>",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/gabliam/gabliam/issues"
  },
  "homepage": "https://github.com/gabliam/gabliam/tree/master/packages/core#readme",
  "engines": {
    "node": ">=8.9.0"
  },
  "gitHead": "ec328fa0b62f474d470f74a2fc9637838f54091e"
}
