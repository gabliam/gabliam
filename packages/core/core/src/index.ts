import 'reflect-metadata';

export {
  decorate,
  inject,
  injectable,
  interfaces as inversifyInterfaces,
  multiInject,
  named,
  optional,
  postConstruct,
  tagged,
  targetName,
  unmanaged,
} from 'inversify';
export * from './common';
export {
  APP_CONFIG,
  CORE_CONFIG,
  INJECT_CONTAINER_KEY,
  VALUE_EXTRACTOR,
} from './constants';
export * from './container';
export * from './decorator';
export * from './errors';
export * from './gabliam';
export * from './gabliam-options';
export * from './gabliam-utils';
export * from './interfaces';
export * from './joi';
export { configResolver, FileLoader, Loader, Resolver } from './loaders';
export * from './metadatas';
export * from './reflection';
export * from './registry';
export * from './value-extractor';
