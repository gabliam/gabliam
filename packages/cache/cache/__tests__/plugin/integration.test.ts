import { Service } from '@gabliam/core';
import {
  CacheManager,
  CachePut,
  CACHE_MANAGER,
  NoOpCache,
  SimpleCacheManager,
} from '../../src/index';
import { CachePluginTest } from './cache-plugin-test';

let appTest: CachePluginTest;

beforeEach(async () => {
  appTest = new CachePluginTest();
});

afterEach(async () => {
  await appTest.destroy();
});

describe('config test', () => {
  it('without config', async () => {
    await appTest.build();
    const cacheManager = appTest.gab.container.get<CacheManager>(CACHE_MANAGER);
    expect(cacheManager).toMatchSnapshot();
  });

  it('with config', async () => {
    appTest.addConf('application.cacheConfig', {
      defaultCache: 'MemoryCache',
      defaultOptionsCache: {
        max: 1,
      },
    });
    await appTest.build();
    const cacheManager = appTest.gab.container.get<CacheManager>(CACHE_MANAGER);
    expect(cacheManager).toMatchSnapshot();
  });

  describe('bad config', () => {
    it('bad defaultCache', async () => {
      appTest.addConf('application.cacheConfig', {
        defaultCache: 'BadCache',
      });
      await expect(appTest.gab.buildAndStart()).rejects.toMatchSnapshot();
    });

    it('bad cacheManager', async () => {
      appTest.addConf('application.cacheConfig', {
        cacheManager: 'BadCacheManager',
      });
      await expect(appTest.gab.buildAndStart()).rejects.toMatchSnapshot();
    });
  });
});

describe('integrations', () => {
  it('test cache', async () => {
    @Service()
    class TestService {
      @CachePut('hi')
      async hi(surname: string, name: string) {
        return `hi ${surname} ${name}`;
      }
    }
    appTest.addClass(TestService);
    appTest.addConf('application.cacheConfig', {
      defaultCache: 'MemoryCache',
      defaultOptionsCache: {
        max: 5,
      },
    });
    await appTest.build();
    const s = appTest.gab.container.get(TestService);
    const cache = appTest.gab.container.get<CacheManager>(CACHE_MANAGER);
    expect(await s.hi('test', 'test')).toMatchSnapshot();
    expect(await s.hi('test', 'test')).toMatchSnapshot();
    expect(await cache.getCache('hi')).toMatchSnapshot();
    expect(await s.hi('test2', 'test2')).toMatchSnapshot();
    expect(await cache.getCache('hi')).toMatchSnapshot();
  });

  it('with mapCache', async () => {
    class CustomCacheManager extends SimpleCacheManager {}

    @Service()
    class TestService {
      @CachePut('hi')
      async hi(surname: string, name: string) {
        return `hi ${surname} ${name}`;
      }
    }
    appTest.addClass(TestService);
    appTest.addConf('application.cacheConfig', {
      cacheManager: CustomCacheManager,
      groups: {
        default: {
          caches: {
            test: {
              cache: 'MemoryCache',
              options: {
                max: 5,
              },
            },
            test2: {
              cache: NoOpCache,
            },
          },
        },
      },
    });
    await appTest.build();
    const s = appTest.gab.container.get(TestService);
    const cache = appTest.gab.container.get<CacheManager>(CACHE_MANAGER);
    expect(await s.hi('test', 'test')).toMatchSnapshot();
    expect(await s.hi('test', 'test')).toMatchSnapshot();
    expect(await cache.getCache('hi')).toMatchSnapshot();
    expect(await s.hi('test2', 'test2')).toMatchSnapshot();
    expect(await cache.getCache('hi')).toMatchSnapshot();
  });
});
