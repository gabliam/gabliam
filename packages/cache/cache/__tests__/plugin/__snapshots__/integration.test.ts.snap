// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`config test bad config bad cacheManager 1`] = `[CacheManagerPgkNotInstalledError:  CacheManager "BadCacheManager" package has not been found installed. Try to install it: npm install BadCacheManager --save or yarn add BadCacheManager]`;

exports[`config test bad config bad defaultCache 1`] = `[CachePgkNotInstalledError:  Cache "BadCache" package has not been found installed. Try to install it: npm install BadCache --save or yarn add BadCache]`;

exports[`config test with config 1`] = `
SimpleCacheManager {
  "defaultCache": [Function],
  "defaultOptionsCache": Object {
    "max": 1,
  },
  "dynamic": true,
  "group": Map {},
  "startedCache": Map {},
}
`;

exports[`config test without config 1`] = `
SimpleCacheManager {
  "defaultCache": [Function],
  "defaultOptionsCache": undefined,
  "dynamic": true,
  "group": Map {},
  "startedCache": Map {},
}
`;

exports[`integrations test cache 1`] = `"hi test test"`;

exports[`integrations test cache 2`] = `"hi test test"`;

exports[`integrations test cache 3`] = `
MemoryCache {
  "name": "hi",
  "options": Object {
    "max": 5,
  },
  "store": LRUCache {
    "allowStale": false,
    "disposeAfter": null,
    "disposed": null,
    "fetchContext": undefined,
    "fetchMethod": null,
    "free": Stack {
      "heap": Uint8Array [
        0,
        0,
        0,
        0,
        0,
      ],
      "length": 0,
    },
    "head": 0,
    "initialFill": 1,
    "keyList": Array [
      "testtest",
      null,
      null,
      null,
      null,
    ],
    "keyMap": Map {
      "testtest" => 0,
    },
    "max": 5,
    "maxSize": 0,
    "next": Uint8Array [
      0,
      0,
      0,
      0,
      0,
    ],
    "noDeleteOnFetchRejection": false,
    "noDeleteOnStaleGet": false,
    "noDisposeOnSet": false,
    "noUpdateTTL": false,
    "prev": Uint8Array [
      0,
      0,
      0,
      0,
      0,
    ],
    "size": 1,
    "sizeCalculation": undefined,
    "tail": 0,
    "ttl": 0,
    "ttlAutopurge": false,
    "ttlResolution": 1,
    "updateAgeOnGet": false,
    "updateAgeOnHas": false,
    "valList": Array [
      "hi test test",
      null,
      null,
      null,
      null,
    ],
  },
}
`;

exports[`integrations test cache 4`] = `"hi test2 test2"`;

exports[`integrations test cache 5`] = `
MemoryCache {
  "name": "hi",
  "options": Object {
    "max": 5,
  },
  "store": LRUCache {
    "allowStale": false,
    "disposeAfter": null,
    "disposed": null,
    "fetchContext": undefined,
    "fetchMethod": null,
    "free": Stack {
      "heap": Uint8Array [
        0,
        0,
        0,
        0,
        0,
      ],
      "length": 0,
    },
    "head": 0,
    "initialFill": 2,
    "keyList": Array [
      "testtest",
      "test2test2",
      null,
      null,
      null,
    ],
    "keyMap": Map {
      "testtest" => 0,
      "test2test2" => 1,
    },
    "max": 5,
    "maxSize": 0,
    "next": Uint8Array [
      1,
      0,
      0,
      0,
      0,
    ],
    "noDeleteOnFetchRejection": false,
    "noDeleteOnStaleGet": false,
    "noDisposeOnSet": false,
    "noUpdateTTL": false,
    "prev": Uint8Array [
      0,
      0,
      0,
      0,
      0,
    ],
    "size": 2,
    "sizeCalculation": undefined,
    "tail": 1,
    "ttl": 0,
    "ttlAutopurge": false,
    "ttlResolution": 1,
    "updateAgeOnGet": false,
    "updateAgeOnHas": false,
    "valList": Array [
      "hi test test",
      "hi test2 test2",
      null,
      null,
      null,
    ],
  },
}
`;

exports[`integrations with mapCache 1`] = `"hi test test"`;

exports[`integrations with mapCache 2`] = `"hi test test"`;

exports[`integrations with mapCache 3`] = `
NoOpCache {
  "name": "hi",
}
`;

exports[`integrations with mapCache 4`] = `"hi test2 test2"`;

exports[`integrations with mapCache 5`] = `
NoOpCache {
  "name": "hi",
}
`;
